### PR TITLE
Remove The Defined Functions List

### PR DESCRIPTION
**Pull request by CodaFi**
Turns out we can do this now! This exposed a *bunch* of cursed behavior, including:

- Closure capture computation order affecting the type of declarations
- `ContextualizeClosures` fighting `RecontextualizeClosures` - which is a thing now
- `Encodable` being forced at some point after `Decodable` failed being a relied-upon behavior to force `CodingKeys` to exist
- Redeclaration checking for multiple primaries affecting the way we typecheck things, in turn affecting the kinds of dependency edges we record depending on the order of primaries at the command line
- Some, but not all, sequences in stdlib that document `Indices` existing for `BidirectionalCollection` bases actually expose `Indices` for mere `Collection`s.  Whoops.

I'd like somebody from stdlib to look over the changes there, and especially give me a hand documenting these things if this is the right fix.

### CODE REVIEW

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
How this also fixed typo correction, I’ll never know...

**Review by slavapestov — APPROVED** [review]

**Comment by CodaFi** [comment]
This exposes a cycle in conformance checking 

```
struct Foo: Collection {
    public typealias Index = Int
    public typealias Element = Int
    
    public var startIndex: Index {
        return 0
    }
    
    public var endIndex: Index {
        return 0
    }
    
    public subscript(index: Index) -> Iterator.Element {
      fatalError();
    }
    
    public func index(after i: Index) -> Index {
        return i + 1
    }
}
```

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test source compatibility

**Comment by CodaFi** [comment]
@swift-ci please test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

**Comment by CodaFi** [comment]
@swift-ci please smoke test macOS platform

File: stdlib/public/core/PrefixWhile.swift [hunk 1]
```diff
@@ -149,6 +149,10 @@ extension LazyPrefixWhileCollection {
   }
 }
 
+extension LazyPrefixWhileSequence where Base: Collection {
+  public typealias Indices = DefaultIndices<LazyPrefixWhileSequence<Base>>
```
Comment (lorentey) [inline stdlib/public/core/PrefixWhile.swift]: Same (non-essential) question as above.

**Review by lorentey — APPROVED** [review]
This looks good! AFAIK, we fully expect these to be defined here, we just relied on the default value for Collection.Indices. (Which we ought not do in the stdlib...)

Do these changes mean that default values for associated types do not always apply in places where they did before? (I.e., are there any source compatibility implications?)

**Comment by lorentey** [comment]
(Luckily, these get documented automatically -- they inherit the generic Indices docs from Collection.)

File: stdlib/public/core/PrefixWhile.swift (hunk 1 above)
Comment (CodaFi) [reply stdlib/public/core/PrefixWhile.swift]: Unfortunately, these extensions seem to have unique bounds.

**Comment by CodaFi** [comment]
There are source compatibility issues even with the current set of workarounds.  I'm going to keep hacking away at this and split the easy stuff into smaller PRs.

**Pull request closed without merging**